keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -73,636 +85,558 @@
+   8. Controller validates BucketClaim and Bucket fields to ensure provisioning success
+   9. Controller copies Bucket status items to BucketClaim status as needed. Importantly:
+      1. Supported protocols
+      2. `ReadyToUse`
+
+#### Deleting a Bucket
+
+User deletes a BucketClaim they no longer need.
+
+Users cannot delete global-scope Buckets directly. Instead, a User deletes a BucketClaim they have have delete permissions for, and COSI coordinates deleting the Bucket.
+
+Each Bucket has a deletion policy that determines whether COSI deletes the Bucket resource when the BucketClaim it's bound to is deleted.
+Deletion policy options:
+- `Delete`: Bucket and underlying OSP bucket are deleted as part of BucketClaim deletion process
+- `Retain`: Bucket is unbound, but the Bucket resource and underlying OSP bucket are kept
+
+An Admin (or other privileged user) can delete any Bucket resource at any time. When a Bucket is deleted, COSI should prevent the Bucket from being deleted until the BucketClaim it is bound to is also in deleting state.
+
+BucketClaims having valid BucketAccesses (i.e., claims in use) will not be deleted until all the BucketAccesses are cleaned up.
+
+When a BucketClaim with Bucket reclaim policy `Reclaim` that is deleted, the Bucket is left in place when the BucketClaim is removed. COSI does not support the Bucket being automatically re-bound to a new BucketClaim without Admin intervention. Admins must be able to
+
+1. User deletes BucketClaim object
+2. COSI Controller detects BucketClaim resource's deletion timestamp
+   1. Controller looks up the corresponding Bucket
+      1. If Bucket doesn't exist, go to **CLEANUP** (Bucket was deleted or never existed)
+      2. If Bucket-BucketClaim binding is not valid, error out
+      3. Apply `objectstorage.k8.io/bucketclaim-being-deleted` annotation to Bucket
+         (tells Sidecar that it's safe to proceed with Bucket deprovisioning)
+      4. If Bucket deletion policy is `Delete`, add deletion timestamp to Bucket
+      5. If `Retain`, nothing more for Controller to do
+      6. **CLEANUP**: Controller removes BucketClaim `objectstorage.k8s.io/bucketclaim-protection` finalizer
+3. COSI Sidecar detects Bucket update
+   1. If the Bucket's driver matches the sidecar's driver, continue
+   2. If `objectstorage.k8.io/bucketclaim-being-deleted` annotation, continue
+   3. If reclaim policy is `Delete`
+      1. If Bucket has nil deletion timestamp, exit (do not deprovision without deletion timestamp)
+      2. Sidecar calls the COSI driver via gRPC to de-provision the OSP bucket
+      3. If OSP returns provision fail, Sidecar reports error to Bucket status and retries gRPC call
+      4. When OSP returns provision success, remove Bucket `objectstorage.k8s.io/bucket-protection` finalizer
+   4. If deletion policy is `Retain`, nothing more to do
+
+COSI Sidecar should not have Bucket delete permissions.
+
+#### Generating Bucket Access Credentials
```

Comment:
Because we are trying to keep the Sidecar responsibilities (and RBAC) as small as possible (following the design of Volume Snapshotter), we needed significant changes to how BucketAccesses are provisioned.

TL;DR: the controller gets info from BucketAccessClass and BucketClaim and copies it to the BucketAccess.status, after which the sidecar can do its thing. Sidecar no longer does Class/Claim inspection.